### PR TITLE
Generate k8s deployment yaml file for cloud service

### LINKED ISSUE #149: Generate cloud service k8s deployment script
#### Software and Hardware Versions 
ModelCI v1.x.x, CUDA Version vx.x.x, GPU device used...

#### Problem description
Generate cloud service deployment yaml file for k8s cluster with template based on current dispachter.

#### Expected Behavior
Generate correct yaml file:
- init container for retrieving model from cloud storage. e.g. S3 bucket
- container with correct image based on model framework and device info
- correct environment variables

#### Other Information 
Jinja2 as the framework for parsing template.

### PR DESCRIPTION
**Pull request by FerdinandZhong**
Closes #149 

#### What has been done to verify that this works as intended?
- pytest to check yaml file can correctly generated
- manually check yaml file items (lack of resources to apply the deployment)

#### Why is this the best possible solution? Were any other approaches considered?
Dynamic generation of cloud service deployment yaml file can help users to easier launch the service into their k8s cluster with correct container settings and dependencies. Verifications of technical details:
- Jinja2: Simplify the parsing of template
- Retrieval of model from cloud storage: Unlike docker, it's hard for k8s pod to bind with physical storage. Pulling model from cloud storage and attach it to the pod is a common methodology. 
  - How to: Pull model through init-container using pre-defined image.
  - Possible other approaches: Provide interface to user to build image with model embedded. Launch k8s pod directly with new image.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Minimum influences to users. As the changes only help generate a deployment file.
However, users are required to provide cloud storage details and other configurations through config file. 

#### Does this change require updates to documentation?
Yes. Sample config file is necessary.

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

**Comment by HuaizhengZhang** [comment]
Thanks for the PR. We will review it soon.

File: modelci/hub/deployer/k8s_dispatcher/sample.config [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[remote_storage]
+# configuration for pulling models from cloud storage.
+storage_type = S3
+aws_access_key_id = SAMPLEKEY
+aws_secret_access_key = SAMPLEACCESSKEY
+bucket_name = ferdinand-own-bucket
+remote_model_path = test-objects/naturallanguageprocessingwithpytorch.pdf
```
Comment (HuaizhengZhang) [inline modelci/hub/deployer/k8s_dispatcher/sample.config:7]: have this PR been tested on k8s?

File: k8s/templates/dispatch_api_template.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  selector:
+    matchLabels:
+        application: ml-model-ci-deployment
+  template:
+    metadata:
+      namespace: {{ deployment.namespace }}
+      labels:
+        application: ml-model-ci-deployment
+    spec:
+      initContainers:
```
Comment (HuaizhengZhang) [inline k8s/templates/dispatch_api_template.yml:20]: So, now this PR only supports model retrieval and dispatch APIs?

File: modelci/hub/deployer/k8s_dispatcher/k8s_dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    deployment_content = template.render(
+        init_container=init_container,
+        init_env_object_list=init_env_object_list,
+        deployment=deployment,
+        env_object_list=env_object_list,
+        port_object_list=port_object_list
+    )
+
+    with open(output_file_path, 'w+') as output_target:
```
Comment (HuaizhengZhang) [inline modelci/hub/deployer/k8s_dispatcher/k8s_dispatcher.py:142]: Can you also update the readme.md under the deployer dir?

This PR contains no example or test cases.

**Comment by HuaizhengZhang** [comment]
Please resolve the conflicts

File: modelci/hub/deployer/k8s_dispatcher/sample.config (hunk 1 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s_dispatcher/sample.config]: yes, tested in minikube with onnx model

File: modelci/hub/deployer/k8s_dispatcher/k8s_dispatcher.py (hunk 3 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s_dispatcher/k8s_dispatcher.py]: Added a readme

File: k8s/templates/dispatch_api_template.yml (hunk 2 above)
Comment (FerdinandZhong) [reply k8s/templates/dispatch_api_template.yml:20]: 1. Model retreval part in deployment script is different from modelci retrieval api.
modelci retrieval api is not suitable for k8s deployment (local storage, stateful service e.g. mongoDB)
therefore I assume use has remote storage of compiled models.
2. This PR integrated with dispatch part. Targeting at providing a deployment script for deploying cloud service with model.

File: modelci/hub/deployer/k8s/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+`[model]` defines variables of model path in containers
+
+`[deployment]` defines variables for serving the model as a cloud service
+
+</li>
+
+<li> Generate deployment file to desired output path.
+
+```
+from modelci.hub.deployer.k8s.k8s_dispatcher import serve
```
Comment (xneng) [inline modelci/hub/deployer/k8s/README.md:44]: Duplicated prefix: `from modelci.hub.deployer.k8s.dispatcher import serve`

File: modelci/hub/deployer/k8s/k8s_dispatcher.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        env_object['value'] = value
+        env_object_list.append(env_object)
+    return env_object_list
+
+def serve(
+        configuration: str = 'config.ini',
+        output_file_path: str = './k8s/deployment.yaml',
+        **additional_environment
+):
+    """Generate the kubernetes deployment yaml file with template.
```
Comment (xneng) [inline modelci/hub/deployer/k8s/k8s_dispatcher.py:39]: I think the serve function's behavior should be creating a k8s workload, currently this function is rendering the k8s config file. It's the pre step of `serve`, so we can rename it like `render`?

**Review by xneng — COMMENTED** [review]
LGFM. And there're some suggestions.

File: modelci/hub/deployer/k8s/README.md (hunk 4 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/README.md:44]: Thanks, will modify this

File: modelci/hub/deployer/k8s/k8s_dispatcher.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+import configparser
+from enum import Enum, unique
+
+from jinja2 import Environment, FileSystemLoader
+
+from modelci.types.bo import Engine
+from modelci.utils.misc import get_device
+
+jinja = Environment(
+    loader=FileSystemLoader('./k8s/templates'),
```
[the remark this answers is not in the document]
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/k8s_dispatcher.py:10]: agree, thanks

File: modelci/hub/deployer/k8s/k8s_dispatcher.py (hunk 5 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/k8s_dispatcher.py:39]: I totally agree with you. `render` is a more proper name. To really do auto serving with k8s cluster, the project may need to integration with CI/CD tools to deploy the rendered file to k8s cluster. I will modify it.

File: modelci/hub/deployer/k8s/k8s_dispatcher.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/k8s_dispatcher.py]: modified

File: modelci/hub/deployer/k8s/k8s_dispatcher.py (hunk 5 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/k8s_dispatcher.py]: modified

File: modelci/hub/deployer/k8s/README.md (hunk 4 above)
Comment (FerdinandZhong) [reply modelci/hub/deployer/k8s/README.md]: modified

**Review by xneng — APPROVED** [review]
Looks good for me.

**Review by HuaizhengZhang — APPROVED** [review]

**Pull request merged**